Update nodes section (#30)
Zilliqa/dev-portal: docs/api-websocket.md

---
id: api-websocket
title: Zilliqa WebSocket Service
---

> 

Reviewer: Need an introduction on what is the purpose for this feature

Reviewer: Need to provide information on the websocket end point link

> This page describes the protocol, between the Zilliqa Websocket Server and the sdk client, for querying subscription and message pushing.

Reviewer 2: ```suggestion
This page describes the protocol - between the Zilliqa Websocket Server and the SDK client - for querying subscription and message pushing.
```

> This page describes the protocol, between the Zilliqa Websocket Server and the sdk client, for querying subscription and message pushing.

Reviewer: Zilliqa Websocket Server (ZWS)

## Feature workflow

> Client can subscribe their interested topics or unsubscribe certain topic by sending query, if the query failed they will normally be informed immediately with related error message. For every Tx block(epoch), the subscribed content will be sent from server to each client in one message where an array contains all their subscribed topic if updated, which we name **notification**.

Reviewer 2: ```suggestion
Clients can subscribe to or unsubscribe from certain topics by sending a query. If the query fails, clients will normally be informed immediately through a relevant error message. For every Tx block, the subscribed content will be sent out by the server to each client in one message. This message (herein referred to as the **notification**) includes an array that contains all the updates to subscribed topics.
```

## Supported query

> The following types of data are the current main focus that we want to consider to be supported by ZWS:

Reviewer 2: ```suggestion
The following types of data are currently supported for querying:
```

> - **New TxBlock**. Which includes TxBlock recently generated and hashes of all the transaction being processed within this block.

Reviewer 2: ```suggestion
- **New TxBlock**. This includes the recently generated Tx block and hashes of all the transactions processed within this block.
```

> - **Event log**. Which includes all the event log generated for interested contract address

Reviewer 2: ```suggestion
- **Event log**. This includes all the event logs generated for the specified contract address.
```

- **Unsubscribe**. Which tells the server to unsubscribe certain topic for the client

## Exception handling

> Usually an **error message** will be responded to the client if the query failed, it may looks like

Reviewer: Usually an **error message** will be sent to the client

```json
{
  "type":"Error",
  "error":"..."
}
```

The following error messages will be applied to all kinds of query if being invalid:

- **invalid query field**. Which tells the client if the query is invalid, it could be not found, empty, malformed, or not available

## Message encoding

For convention, we still use JSON as our encoding format.

> The epoch message will be presented in this way:

Reviewer 3: yes

					"NumMicroBlocks": 3,
					"NumTxns": 1,
					"PrevBlockHash": "5bda21605e7ea9404c58a40eebe99563adf330bab5b39e7438f8e4db28607b37",
					"Rewards": "1000000000",
					"StateDeltaHash": "2f878030ab9b0a211c1e584e140707c79d62d067390bfe3ccaf08fdaeaad2229",
					"StateRootHash": "94abb63e27984f46e914db2601de6af2048a3bf72f69eaac34421b7dfbd34a82",
					"Timestamp": "1572512230807870",
					"Version": 1
				}
			}
		},
		"TxHashes": [
			["1beb32a5435e993aa3025a70d8a5e71df43c10e2fe3f6ef832d1a5c371a63852"],
			[],
			[]
		]
	}, {
		"query": "EventLog",
		"value": [{
			"address": "0x0000000000000000000000000000000000000000",
			"event_logs": [{
				"_eventname": "foo1",
				"params": [{
					"vname": "bar1",
					"type": "String",
					"value": "abc"
				}, {
					"vname": "bar2",
					"type": "ByStr32",
					"value": "0x0000000000000000000000000000000000000001"
				}]
			}]
		}]
	}, {
		"query": "Unsubscribe",
		"value": ["NewBlock"]
	}]
}
```

The followings are case by case for each subscription:

### Subscribe New Block

#### query message

```json
{
	"query": "NewBlock"
}
```

#### response message

> Once succsfully subscribed, server will echo the query message to the client, otherwise will return error message.

Reviewer: successfully

					"MicroBlockShardId": 2,
					"MicroBlockTxnRootHash": "0000000000000000000000000000000000000000000000000000000000000000"
				}]
			},
			"header": {
				"BlockNum": "15",
				"DSBlockNum": "1",
				"GasLimit": "15000000",
				"GasUsed": "1",
				"MbInfoHash": "4b2d20a0bcb382ad2e2560791ed90ed21100e8e84ebac63d62d3c0b1a3fb11fe",
				"MinerPubKey": "0x02FC9ED69524A23AEFCB85B8A36C998F512C0512C6932DED74680A044F9D3EBC95",
				"NumMicroBlocks": 3,
				"NumTxns": 1,
				"PrevBlockHash": "5bda21605e7ea9404c58a40eebe99563adf330bab5b39e7438f8e4db28607b37",
				"Rewards": "1000000000",
				"StateDeltaHash": "2f878030ab9b0a211c1e584e140707c79d62d067390bfe3ccaf08fdaeaad2229",
				"StateRootHash": "94abb63e27984f46e914db2601de6af2048a3bf72f69eaac34421b7dfbd34a82",
				"Timestamp": "1572512230807870",
				"Version": 1
			}
		}
	},
	"TxHashes": [
		["1beb32a5435e993aa3025a70d8a5e71df43c10e2fe3f6ef832d1a5c371a63852"],
		[],
		[]
	]
}
```

### Subscribe Event Log

#### query message

```json
{
  "query":"EventLog",
  "addresses":[
    "0x0000000000000000000000000000000000000000",
    "0x1111111111111111111111111111111111111111"
  ]
}
```

#### response message

> Once succesfully subscribed, server will echo the query message to the client,

Reviewer: successfully

otherwise will return error message.

Special error message:

- **invalid addresses field**, which tells the client the addresses field is invalid, it could either be not found, malformed or empty
- **no contract found in list**, which tells the client the addresses provided are all of non contract

#### expected field in notification

```json
{
  "query":"EventLog",
  "value":
  [
    {
      "address":"0x0000000000000000000000000000000000000000",
      "event_logs":[
        {
          "_eventname":"foo1",
          "params":[
            {
              "vname":"bar1",
              "type":"String",
              "value":"abc"
            },
            {
              "vname":"bar2",
              "type":"ByStr32",
              "value":"0x0000000000000000000000000000000000000001"
            }
          ]
        },
      ]
    }
  ]
}
```

Notice that for address `0x1111111111111111111111111111111111111111` is not presented in the message since it doesn't have any event log released in this epoch.

### Unsubscribe

#### query message

```json
{
  "query":"Unsubscribe",
  "type":"EventLog"
}
```

#### response message

> Once succesfully ubsubscribed, server will echo the query message to the client,

Reviewer 2: Once successfully unsubscribed, the server will echo the query message to the client. Otherwise, the server will return an error message.

---
id: api-websocket

> title: Zilliqa Websocket Server (ZWS)

Reviewer 2: ```suggestion
title: Zilliqa WebSocket Server (ZWS)
```

---

> This page describes the protocol - between the Zilliqa Websocket Server and the SDK client - for querying subscription and message pushing.

Reviewer 2: ```suggestion
This page describes the protocol - between the Zilliqa WebSocket Server and the SDK client - for querying subscription and message pushing.
```

---
id: api-websocket
title: Zilliqa Websocket Server (ZWS)
---

This page describes the protocol - between the Zilliqa Websocket Server and the SDK client - for querying subscription and message pushing.

## Introduction

> Zilliqa WebSocket Server provides WebSocket service, enabling dApp developers or other builders on Zilliqa blockchain to subscribe to specific topics (e.g., new block has been produced, payment transaction has been confirmed, events issued by smart contracts). With the WebSocket service, developers need not to use polling the blockchain routinely in order to get specific information.

Reviewer 2: ```suggestion
Zilliqa WebSocket Server provides WebSocket service, enabling dApp developers or other builders on Zilliqa blockchain to subscribe to specific topics (e.g., new block has been produced, payment transaction has been confirmed, events issued by smart contracts). With the WebSocket service, developers no longer need to poll the blockchain routinely in order to get specific information.
```
